### PR TITLE
FRW-458: Authorization code flow 

### PR DESCRIPTION
**Pull request by olhalivitchuk**
- Developer(s): @olhalivitchuk

- Ticket: https://spryker.atlassian.net/browse/FRW-458

- Release Group: https://release.spryker.com/release-groups/view/4509

- PR Overview: https://release.spryker.com/release/pull-request-eco/AuthorizationPickingAppBackendApi/1

- merge: squash

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   AuthorizationPickingAppBackendApi                 |   minor (new)              |    Oauth, OauthCodeFlow                   |

-----------------------------------------

##### Change log

Initial Release

- Introduced `AuthorizationPickingAppBackendApi` module that allows `PickingApp` to authorize via `AuthorizationCode` flow.

### CODE REVIEW

**Review by asmarovydlo — CHANGES_REQUESTED** [review]

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    "description": "AuthorizationPickingAppBackendApi module",
+    "license": "MIT",
+    "require": {
+        "php": ">=8.0",
+        "psr/http-message": "^1.0.0",
+        "spryker/glue-application-extension": "^1.10.0",
+        "spryker/guzzle": "^2.0.0",
+        "spryker/kernel": "^3.52.0",
+        "spryker/oauth": "^2.6.0",
+        "spryker/propel-orm": "^1.16.0",
```
[the remark this answers is not in the document]
Comment (olhalivitchuk) [reply composer.json:13]: Is it needed? Our dependency finder tool resolves it as it mapped external package `Bundles/Development/src/Spryker/Zed/Development/DevelopmentConfig.php`

[the remark this answers is not in the document]
Comment (asmarovydlo) [reply composer.json:13]: 👍

**Review by asmarovydlo — CHANGES_REQUESTED** [review]
Fixed only parially https://github.com/spryker-eco/authorization-picking-app-backend-api/pull/1#discussion_r1031731646

**Review by asmarovydlo — APPROVED** [review]

**Comment by asmarovydlo** [comment]
**[Note for releaser]** Travis is red because `composer.json` pointed to not yet released version of `spryker/oauth` module. (It was added in the last commit, the previous build is green)

**Comment by dereuromark** [comment]
Please dont use invalid labels, use the ones release app assigns to the repo.

**Comment by dereuromark** [comment]
> Warning: Module AuthorizationPickingAppBackendApi has PHPStan check level 7, expected 8 for major or new module.

**Comment by asmarovydlo** [comment]
> Please dont use invalid labels, use the ones release app assigns to the repo.
Sorry, didn't have this label in the labels list. How you managed to add it? @dereuromark

**Comment by asmarovydlo** [comment]
> > Warning: Module AuthorizationPickingAppBackendApi has PHPStan check level 7, expected 8 for major or new module.

It is unstable 0.x module. Does it make sense?

**Comment by dereuromark** [comment]
ReleaseApp can do it, you just need to talk to me.

Well, 0.x is not that critical, indeed. Still always make sense as it avoids errors.

**Review by asmarovydlo — APPROVED** [review]

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    },
+    "config": {
+        "sort-packages": true,
+        "process-timeout": 600,
+        "allow-plugins": {
+            "dealerdirect/phpcodesniffer-composer-installer": true
+        }
+    },
+    "include-path": [
+        "src/"
```
Comment (dereuromark) [inline composer.json:69]: What is this? Why is this needed?

File: phpstan.neon [hunk 3]
```diff
@@ -0,0 +1,8 @@
+parameters:
+    level: 7
+    bootstrapFiles:
+       - tests/bootstrap.php
+    checkGenericClassInNonGenericObjectType: false
+    checkMissingIterableValueType: false
+    ignoreErrors:
+            - '#Call to .+ on an unknown class Generated.+AutoCompletion#'
```
Comment (dereuromark) [inline phpstan.neon:8]: This can get fixed with proper test harness as per our docs!

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                  composer --version
+                  composer install
+            - name: Run codestyle checks
+              run: composer cs-check
+
+            - name: Run codeception tests (failing because of missing `spryker/oauth` dependency)
+              run: composer test || true
+
+            - name: Run PHPStan (failing because of missing `spryker/oauth` dependency)
+              run: composer stan || true
```
Comment (dereuromark) [inline .github/workflows/ci.yml:88]: Is there a ticket to fix this after release?

Comment (olhalivitchuk) [reply .github/workflows/ci.yml:88]: https://spryker.atlassian.net/browse/FRW-1101

**Comment by asmarovydlo** [comment]
> The module does not meet the standards for a standalone module yet Please also make sure it gets a prefer lowest check as per our docs and guidelines, either before or after the release This is especially important due to the many dependencies it has.

Thanks for the feedback!
How critical is it taking into account that the module has 0.x version?

**Comment by dereuromark** [comment]
This is not related to the 0.x version per se, but the "truth" of the dependencies in all "promised" versions.
But sure, it can be a follow up once the tests are working again, before that it doesnt make too much sense anyway.

File: composer.json (hunk 2 above)
Comment (dereuromark) [reply composer.json:69]: This deprecated property is not allowed in our spryker code
https://getcomposer.org/doc/04-schema.md#include-path

You will need to prove the need for it IMO.

**Comment by olhalivitchuk** [comment]
Collected all comments in follow-up ticket https://spryker.atlassian.net/browse/FRW-1101

File: composer.json (hunk 2 above)
Comment (dereuromark) [reply composer.json:69]: @olhalivitchuk This is still a blocker for release. Can you please check in a separate PR that this can be safely removed?

Comment (dereuromark) [reply composer.json]: You just need to make sure helpers are in psr4 autoload setup. Then it should work fine for all classes as usual.

Comment (olhalivitchuk) [reply composer.json]: yep, it works

File: phpstan.neon (hunk 3 above)
Comment (dereuromark) [reply phpstan.neon:8]: https://spryker.atlassian.net/wiki/spaces/CORE/pages/2868248862/Standalone+module+testing
and as per most other standalone modules that have a valid test harness set up

Comment (olhalivitchuk) [reply phpstan.neon:8]: Based on this doc there some example modules with valid settings https://spryker.atlassian.net/wiki/spaces/CORE/pages/2868248862/Standalone+module+testing#Example-integrations

but all of them have `Call to .+ on an unknown class Generated.+AutoCompletion` in `phpstan.neon`

Comment (dereuromark) [reply phpstan.neon:8]: Those must be the outdated onces
There are at least several that showcase the valid setup
https://github.com/spryker-eco/loggly

Comment (olhalivitchuk) [reply phpstan.neon:8]: Thanks!
But `Loggly` does not have `DependencyProvider`, so that's the reason it does not have such ignored error. 
We get the error for DP, not for tests

Comment (dereuromark) [reply phpstan.neon:8]: Hm, I am 100% sure there is one, because I personally fixed it at the time
Running the IDE generation itself would solve this afaik.
